mkoctfile / g++ is unable to compile and link oct files in confined snap environment

This is the biggest blocker to normal functioning of this snap. The gcc and g++ that are included in the confined snap environment run, but they are unable to include header files and link with libraries in their standard locations.

The initial error comes from the `-I` include search options that were used to build Octave in the snapcraft build environment. The compiler options used to build Octave are stored and applied when the user wants to build an oct file extension. This normally works in a standard system.

In the snap confined environment, the user gets this

```
>> mkoctfile example.cc
cc1plus: error: /root/parts/octave/install/usr/include: Permission denied
cc1plus: error: /root/parts/octave/install/usr/include/x86_64-linux-gnu: Permission denied
cc1plus: error: /root/parts/octave/install/usr/include: Permission denied
cc1plus: error: /root/parts/octave/install/usr/include/x86_64-linux-gnu: Permission denied
warning: mkoctfile: building exited with failure status
```

Obviously all of these /root/parts/octave directories are relevant only to the snapcraft build environment.

Fortunately, all of these options can be overridden. So let's say the snap is updated to override all the necessary compiler options with the following

```
CFLAGS=-O2
CXXFLAGS=-O2
CPPFLAGS="-isystem $SNAP/usr/include/c++/7 -isystem $SNAP/usr/include/x86_64-linux-gnu -isystem $SNAP/usr/include"
```

The user now gets back a link error about missing core GCC libraries

```
>> mkoctfile foo.cc
/snap/octave/5/usr/bin/ld: cannot find /usr/lib/x86_64-linux-gnu/libmvec_nonshared.a
collect2: error: ld returned 1 exit status
warning: mkoctfile: building exited with failure status
```

I haven't been able to get past this error, because at this point, we have _some_ of the GCC system libraries in /lib and /usr/lib (from the core18 base), but the rest of the development libraries in $SNAP/usr/lib, for example $SNAP/usr/lib/x86_64-linux-gnu/libmvec_nonshared.a. But libm.so is a linker script which specifically mentions /usr/lib/x86_64-linux-gnu/libmvec_nonshared.a by its full path.

We could try to rewrite any linker scripts which are installed by libc6-dev as a stage-package into Octave's $SNAP/usr/lib directory. I have a feeling this won't be the last error.

When I was working on this harder about 4 months ago, I had the feeling that this would be much easier if there were something like a `core18-dev` base snap that we could rely on to provide the same functionality as `core18`, with gcc, g++, and libc6-dev included.

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -92,6 +92,9 @@ parts:
       # Strip debug symbols from programs and libraries when installing
       - INSTALL_PROGRAM: "/usr/bin/install -s"
       - INSTALL_STRIP_FLAG: "-s"
+      # Force Octave header files to be found before system header files
+      # This override can be dropped after upgrading to snapcraft version 3.11
+      - INCLUDES: "-Ilibinterp/corefcn"
     override-build: |
       # Patch mkoctfile to record the original build flags from above, so that
       # additional flags from snapcraft build setup aren't leaked

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -163,9 +163,9 @@ parts:
       - libglpk40
       - libglu1-mesa
       - libgraphicsmagick++-q16-12
-      - libhdf5-100
-      - libopenblas-base
-      - libpcre3
+      - libhdf5-dev
+      - libopenblas-dev
+      - libpcre3-dev
       - libportaudio2
       - libqhull7
       - libqrupdate1

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -216,3 +216,17 @@ layout:
   # by qt in 18.04, may not be needed in 20.04
   /usr/share/X11:
     symlink: $SNAP/usr/share/X11
+
+  # Make all system and staged development library header files available in
+  # standard /usr/include location, to allow compiling oct files and packages
+  /usr/include:
+    bind: $SNAP/usr/include
+
+  # Make certain libc static library fragments available in expected locations,
+  # to allow linking oct files and packages
+  /usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libc_nonshared.a:
+    symlink: $SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libc_nonshared.a
+  /usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libmvec_nonshared.a:
+    symlink: $SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libmvec_nonshared.a
+  /usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libpthread_nonshared.a:
+    symlink: $SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/libpthread_nonshared.a

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -112,6 +112,8 @@ parts:
       snapcraftctl prime
       sed -i 's|^Icon=octave|Icon=/share/icons/hicolor/scalable/apps/octave.svg|' share/applications/org.octave.Octave.desktop
       sed -i 's|^prefix=.*|prefix=/usr|' usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig/*.pc
+      rm -f usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig/hdf5.pc
+      ln -s hdf5-serial.pc usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig/hdf5.pc
     prime:
       - -usr/share/texmf/ls-R
     organize:

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -23,6 +23,8 @@ apps:
       OCTAVE_HOME: "$SNAP"
       LD_LIBRARY_PATH: "$SNAP/lib/octave/$SNAPCRAFT_PROJECT_VERSION"
       LOCPATH: "$SNAP/usr/lib/locale"
+      PKG_CONFIG_PATH: "$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig"
+      PKG_CONFIG_SYSROOT_DIR: "$SNAP"
     plugs:
       - desktop
       - desktop-legacy
@@ -39,6 +41,8 @@ apps:
       OCTAVE_HOME: "$SNAP"
       LD_LIBRARY_PATH: "$SNAP/lib/octave/$SNAPCRAFT_PROJECT_VERSION"
       LOCPATH: "$SNAP/usr/lib/locale"
+      PKG_CONFIG_PATH: "$SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig"
+      PKG_CONFIG_SYSROOT_DIR: "$SNAP"
     plugs:
       - desktop
       - desktop-legacy
@@ -107,6 +111,7 @@ parts:
     override-prime: |
       snapcraftctl prime
       sed -i 's|^Icon=octave|Icon=/share/icons/hicolor/scalable/apps/octave.svg|' share/applications/org.octave.Octave.desktop
+      sed -i 's|^prefix=.*|prefix=/usr|' usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pkgconfig/*.pc
     prime:
       - -usr/share/texmf/ls-R
     organize:
@@ -135,6 +140,7 @@ parts:
       - libreadline-dev
       - libsndfile1-dev
       - libsuitesparse-dev
+      - pkg-config
       - portaudio19-dev
       - qtbase5-dev
       - qttools5-dev
@@ -183,6 +189,7 @@ parts:
       - libsndfile1
       - libumfpack5
       - locales-all
+      - pkg-config
       - texinfo
 
 layout:

